### PR TITLE
[PATCH API-NEXT v3] Asynchronous ordered context lock

### PR DESCRIPTION
**Pull request by bala-manoharan**
add asynchronous ordered context lock.

### CODE REVIEW

File: include/odp/api/spec/schedule.h [hunk 1]
```diff
@@ -371,6 +371,30 @@ void odp_schedule_order_unlock(uint32_t lock_index);
  */
 void odp_schedule_order_unlock_lock(uint32_t unlock_index, uint32_t lock_index);
 
+/** Asynchronous ordered context lock
+ * Request an ordered context lock to be acquired. Starts to acquire an ordered
+ * context lock does not wait until the lock has been acquired. Application can
```
Comment (psavol) [inline include/odp/api/spec/schedule.h]: Suggesting re-phrase the sentence: "Starts an ordered context lock acquire operation, but does not wait until..."

File: include/odp/api/spec/schedule.h [hunk 2]
```diff
@@ -371,6 +371,30 @@ void odp_schedule_order_unlock(uint32_t lock_index);
...
+ *			- 1.
+ *
+ */
+void odp_schedule_order_lock_start(uint32_t lock_index);
+
+/** Asynchronous ordered context lock wait
+ * wait for a previously started lock acquire operation to finish.
+ * Lock index must match with the previous start call. Ordered lock acquisition
+ * will be completed during this call.
+ */
```
Comment (psavol) [inline include/odp/api/spec/schedule.h]: @param lock_index missing

File: include/odp/api/spec/schedule.h [hunk 3]
```diff
@@ -371,6 +371,30 @@ void odp_schedule_order_unlock(uint32_t lock_index);
...
 void odp_schedule_order_unlock_lock(uint32_t unlock_index, uint32_t lock_index);
 
+/** Asynchronous ordered context lock
+ * Request an ordered context lock to be acquired. Starts to acquire an ordered
+ * context lock does not wait until the lock has been acquired. Application can
+ * use this call to potentially interleave some processing within waiting for
+ * this lock. Each start lock call must be paired with a wait call that blocks
+ * until the lock has been acquired. Locks cannot be acquired in nested fashion
+ * i.e each start call must follow a paring wait call before the next ordered
+ * context lock call (both sync and async).
```
Comment (psavol) [inline include/odp/api/spec/schedule.h]: Terms "sync and async" are not needed, we can just say:

"... must follow paring wait and unlock calls, before using another lock."

File: include/odp/api/spec/schedule.h [hunk 4]
```diff
@@ -371,6 +371,30 @@ void odp_schedule_order_unlock(uint32_t lock_index);
...
+ *
+ * @param lock_index	Index of the ordered lock in the current context to be
+ *			acquired. Must be in the range 0..odp_queue_lock_count()
+ *			- 1.
+ *
+ */
+void odp_schedule_order_lock_start(uint32_t lock_index);
+
+/** Asynchronous ordered context lock wait
+ * wait for a previously started lock acquire operation to finish.
```
Comment (psavol) [inline include/odp/api/spec/schedule.h]: Upper case W: "Wait"

File: include/odp/api/spec/schedule.h [hunk 5]
```diff
@@ -371,6 +371,30 @@ void odp_schedule_order_unlock(uint32_t lock_index);
...
+ * use this call to potentially interleave some processing within waiting for
+ * this lock. Each start lock call must be paired with a wait call that blocks
+ * until the lock has been acquired. Locks cannot be acquired in nested fashion
+ * i.e each start call must follow a paring wait call before the next ordered
+ * context lock call (both sync and async).
+ * The same constraints apply as with odp_schedule_order_lock()
+ *
+ * @param lock_index	Index of the ordered lock in the current context to be
+ *			acquired. Must be in the range 0..odp_queue_lock_count()
+ *			- 1.
```
Comment (Bill-Fischofer-Linaro) [inline include/odp/api/spec/schedule.h]: Extraneous `-1` here?

Comment (Bill-Fischofer-Linaro) [inline include/odp/api/spec/schedule.h]: I'd reflow this line so it doesn't look like an extraneous `-1` got inserted here.

**Comment by bala-manoharan** [comment]
V3 incorporates review comments from V2

**Comment by muvarov** [comment]
Merged.

**Pull request closed without merging**